File: application/amplify/functions/evaluation-runner/handler.ts
import type { Handler, EventBridgeEvent, APIGatewayProxyEvent } from 'aws-lambda';
import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import {
  DynamoDBDocumentClient,
  GetCommand,
  UpdateCommand,
  QueryCommand,
  ScanCommand
} from '@aws-sdk/lib-dynamodb';
import { BedrockRuntimeClient, ConverseCommand } from '@aws-sdk/client-bedrock-runtime';
import { S3Client, GetObjectCommand } from '@aws-sdk/client-s3';

// W&B SDK import (using dynamic import for CommonJS compatibility)
let wandb: typeof import('wandb') | null = null;

const dynamoClient = new DynamoDBClient({});
const docClient = DynamoDBDocumentClient.from(dynamoClient);
const bedrockClient = new BedrockRuntimeClient({});
const s3Client = new S3Client({});

// Environment variables
const WANDB_API_KEY = process.env.WANDB_API_KEY;
const WANDB_PROJECT = process.env.WANDB_PROJECT || 'biz-doc-vqa';
const WANDB_ENTITY = process.env.WANDB_ENTITY;

// Table names (passed via environment)
const EVALUATION_JOB_TABLE = process.env.EVALUATION_JOB_TABLE_NAME || '';
const ANNOTATION_TABLE = process.env.ANNOTATION_TABLE_NAME || '';
const IMAGE_TABLE = process.env.IMAGE_TABLE_NAME || '';
const STORAGE_BUCKET = process.env.STORAGE_BUCKET_NAME || '';

// Model configurations
const MODEL_CONFIGS: Record<string, { modelId: string; displayName: string }> = {
  'Amazon Nova Pro': { modelId: 'amazon.nova-pro-v1:0', displayName: 'Amazon Nova Pro' },
  'Claude 3.5 Sonnet': { modelId: 'anthropic.claude-3-5-sonnet-20241022-v2:0', displayName: 'Claude 3.5 Sonnet' },
  'Nemotron Nano 12B': { modelId: 'nvidia.nemotron-nano-12b', displayName: 'Nemotron Nano 12B' }
};

interface EvaluationEvent {
  jobId?: string;
  modelName?: string;
  datasetVersion?: string;
  triggeredBy?: string;
  triggerType?: 'SCHEDULED' | 'MANUAL';
}

interface AnnotationData {
  id: string;
  imageId: string;
  question: string;
  answer: string;
  language: string;
  boundingBoxes: number[][];
  questionType: string;
}

interface ImageData {
  id: string;
  fileName: string;
  s3Key: string;
  width: number;
  height: number;
}

interface EvaluationResult {
  annotationId: string;
  predicted: string;
  groundTruth: string;
  exactMatch: number;
  f1Score: number;
  iou: number;
  predictedBbox?: number[];
}

async function initWandb(): Promise<void> {
  if (!wandb) {
    wandb = await import('wandb');
  }
}

async function updateJobStatus(
  jobId: string,
  status: string,
  updates: Record<string, unknown>
): Promise<void> {
  const updateParts: string[] = [];
  const expressionValues: Record<string, unknown> = {};

  Object.entries({ status, ...updates }).forEach(([key, value]) => {
    if (value !== undefined) {
      updateParts.push(`${key} = :${key}`);
      expressionValues[`:${key}`] = value;
    }
  });

  try {
    await docClient.send(new UpdateCommand({
      TableName: EVALUATION_JOB_TABLE,
      Key: { id: jobId },
      UpdateExpression: `SET ${updateParts.join(', ')}`,
      ExpressionAttributeValues: expressionValues
    }));
  } catch (error) {
    console.error(`Failed to update job ${jobId}:`, error);
    throw error;
  }
}

async function getApprovedAnnotations(): Promise<AnnotationData[]> {
  try {
    const result = await docClient.send(new ScanCommand({
      TableName: ANNOTATION_TABLE,
      FilterExpression: 'validationStatus = :status',
      ExpressionAttributeValues: {
        ':status': 'APPROVED'
      }
    }));
    return result.Items as AnnotationData[] || [];
  } catch (error) {
    console.error('Failed to get approved annotations:', error);
    return [];
  }
}

async function getImage(imageId: string): Promise<ImageData | null> {
  try {
    const result = await docClient.send(new GetCommand({
      TableName: IMAGE_TABLE,
      Key: { id: imageId }
    }));
    return result.Item as ImageData | null;
  } catch (error) {
    console.error(`Failed to get image ${imageId}:`, error);
    return null;
  }
}

async function getImageFromS3(s3Key: string): Promise<Buffer | null> {
  try {
    const response = await s3Client.send(new GetObjectCommand({
      Bucket: STORAGE_BUCKET,
      Key: s3Key
    }));
    const bytes = await response.Body?.transformToByteArray();
    return bytes ? Buffer.from(bytes) : null;
  } catch (error) {
    console.error(`Failed to get image from S3 ${s3Key}:`, error);
    return null;
  }
}

async function callModel(
  modelId: string,
  imageBuffer: Buffer,
  question: string
): Promise<{ answer: string; bbox?: number[] }> {
  const prompt = `You are evaluating an OCR model. Look at this business document image and answer the following question.

Question: ${question}

Return your response as JSON with this format:
{
  "answer": "the extracted answer from the document",
  "boundingBox": [x0, y0, x1, y1]  // normalized coordinates 0-1
}

If you cannot find the answer, return {"answer": "", "boundingBox": null}`;

  try {
    const command = new ConverseCommand({
      modelId,
      messages: [{
        role: 'user',
        content: [
          { text: prompt },
          {
            image: {
              format: 'jpeg',
              source: { bytes: imageBuffer }
            }
          }
        ]
      }],
      inferenceConfig: {
        temperature: 0.1,
        maxTokens: 500
      }
    });

    const response = await bedrockClient.send(command);
    const responseText = response.output?.message?.content?.[0]?.text || '';

    // Try to parse JSON response
    try {
      const parsed = JSON.parse(responseText);
      return {
        answer: parsed.answer || '',
        bbox: parsed.boundingBox
      };
    } catch {
      // If JSON parsing fails, return raw text
      return { answer: responseText };
    }
  } catch (error) {
    console.error('Model call failed:', error);
    return { answer: '' };
  }
}

function calculateExactMatch(predicted: string, groundTruth: string): number {
  const predNorm = predicted.trim().toLowerCase();
  const gtNorm = groundTruth.trim().toLowerCase();
  return predNorm === gtNorm ? 1.0 : 0.0;
}

function calculateF1Score(predicted: string, groundTruth: string): number {
  const predTokens = new Set(predicted.toLowerCase().split(/\s+/).filter(t => t.length > 0));
  const gtTokens = new Set(groundTruth.toLowerCase().split(/\s+/).filter(t => t.length > 0));

  if (predTokens.size === 0 && gtTokens.size === 0) return 1.0;

  const intersection = new Set([...predTokens].filter(t => gtTokens.has(t)));
  if (intersection.size === 0) return 0.0;

  const precision = intersection.size / predTokens.size;
  const recall = intersection.size / gtTokens.size;

  return 2 * (precision * recall) / (precision + recall);
}

function calculateIoU(bbox1?: number[], bbox2?: number[][]): number {
  if (!bbox1 || !bbox2 || bbox2.length === 0) return 0.0;

  // Use the first ground truth bounding box
  const gtBbox = bbox2[0];
  if (gtBbox.length < 4 || bbox1.length < 4) return 0.0;

  const [x1Min, y1Min, x1Max, y1Max] = bbox1;
  const [x2Min, y2Min, x2Max, y2Max] = gtBbox;

  // Calculate intersection
  const xInterMin = Math.max(x1Min, x2Min);
  const yInterMin = Math.max(y1Min, y2Min);
  const xInterMax = Math.min(x1Max, x2Max);
  const yInterMax = Math.min(y1Max, y2Max);

  if (xInterMax <= xInterMin || yInterMax <= yInterMin) return 0.0;

  const interArea = (xInterMax - xInterMin) * (yInterMax - yInterMin);

  // Calculate union
  const area1 = (x1Max - x1Min) * (y1Max - y1Min);
  const area2 = (x2Max - x2Min) * (y2Max - y2Min);
  const unionArea = area1 + area2 - interArea;

  return unionArea > 0 ? interArea / unionArea : 0.0;
}

async function runEvaluation(
  modelName: string,
  datasetVersion: string,
  jobId: string
): Promise<{
  results: EvaluationResult[];
  exactMatchRate: number;
  f1Score: number;
  avgIoU: number;
  wandbRunUrl: string;
}> {
  await initWandb();
  if (!wandb) {
    throw new Error('Failed to initialize W&B SDK');
  }

  const modelConfig = MODEL_CONFIGS[modelName];
  if (!modelConfig) {
    throw new Error(`Unknown model: ${modelName}`);
  }

  // Get approved annotations
  const annotations = await getApprovedAnnotations();
  if (annotations.length === 0) {
    throw new Error('No approved annotations found for evaluation');
  }

  console.log(`Starting evaluation of ${modelName} on ${annotations.length} annotations`);

  // Initialize W&B run
  const run = await wandb.init({
    project: WANDB_PROJECT,
    entity: WANDB_ENTITY || undefined,
    name: `eval-${modelName.toLowerCase().replace(/\s+/g, '-')}-${jobId}`,
    tags: ['evaluation', modelName, datasetVersion],
    config: {
      model: modelName,
      modelId: modelConfig.modelId,
      datasetVersion,
      jobId,
      sampleCount: annotations.length,
      timestamp: new Date().toISOString()
    }
  });

  // Create evaluation artifact
  const evalArtifact = new wandb.Artifact(
    `${modelName.toLowerCase().replace(/\s+/g, '-')}-eval`,
    'evaluation',
    {
      description: `Evaluation of ${modelName} on VQA dataset`,
      metadata: {
        model: modelName,
        dataset_version: datasetVersion,
        evaluated_date: new Date().toISOString()
      }
    }
  );

  // Create results table
  const resultsTable = new wandb.Table({
    columns: ['annotation_id', 'predicted', 'ground_truth', 'exact_match', 'f1_score', 'iou']
  });

  const results: EvaluationResult[] = [];
  let totalEM = 0;
  let totalF1 = 0;
  let totalIoU = 0;

  // Evaluate each annotation
  for (let i = 0; i < annotations.length; i++) {
    const annotation = annotations[i];

    try {
      // Get image
      const image = await getImage(annotation.imageId);
      if (!image) {
        console.warn(`Image ${annotation.imageId} not found, skipping`);
        continue;
      }

      const imageBuffer = await getImageFromS3(image.s3Key);
      if (!imageBuffer) {
        console.warn(`Failed to load image from S3, skipping`);
        continue;
      }

      // Call model
      const prediction = await callModel(modelConfig.modelId, imageBuffer, annotation.question);

      // Calculate metrics
      const em = calculateExactMatch(prediction.answer, annotation.answer);
      const f1 = calculateF1Score(prediction.answer, annotation.answer);

      // Normalize ground truth bbox for IoU calculation (if available)
      let normalizedGtBbox: number[][] | undefined;
      if (annotation.boundingBoxes && annotation.boundingBoxes.length > 0) {
        normalizedGtBbox = annotation.boundingBoxes.map(bbox => [
          bbox[0] / image.width,
          bbox[1] / image.height,
          bbox[2] / image.width,
          bbox[3] / image.height
        ]);
      }

      const iou = calculateIoU(prediction.bbox, normalizedGtBbox);

      // Update running totals
      totalEM += em;
      totalF1 += f1;
      totalIoU += iou;

      const result: EvaluationResult = {
        annotationId: annotation.id,
        predicted: prediction.answer,
        groundTruth: annotation.answer,
        exactMatch: em,
        f1Score: f1,
        iou,
        predictedBbox: prediction.bbox
      };
      results.push(result);

      // Add to table
      resultsTable.addData(
        annotation.id,
        prediction.answer,
        annotation.answer,
        em,
        f1,
        iou
      );

      // Log incremental progress
      run.log({
        'progress/samples_evaluated': i + 1,
        'progress/running_exact_match': totalEM / (i + 1),
        'progress/running_f1_score': totalF1 / (i + 1),
        'progress/running_iou': totalIoU / (i + 1),
        'sample/exact_match': em,
        'sample/f1_score': f1,
        'sample/iou': iou
      });

    } catch (error) {
      console.error(`Error evaluating annotation ${annotation.id}:`, error);
    }
  }

  // Calculate final metrics
  const sampleCount = results.length;
  const exactMatchRate = sampleCount > 0 ? totalEM / sampleCount : 0;
  const avgF1 = sampleCount > 0 ? totalF1 / sampleCount : 0;
  const avgIoU = sampleCount > 0 ? totalIoU / sampleCount : 0;

  // Add results to artifact
  evalArtifact.add(resultsTable, 'evaluation_results');
  run.logArtifact(evalArtifact);

  // Log to run history
  run.log({ evaluation_results: resultsTable });

  // Log summary
  wandb.summary['exact_match_rate'] = exactMatchRate;
  wandb.summary['avg_f1_score'] = avgF1;
  wandb.summary['avg_iou'] = avgIoU;
  wandb.summary['model_name'] = modelName;
  wandb.summary['samples_evaluated'] = sampleCount;

  await evalArtifact.wait();
  const wandbRunUrl = run.url;

  await run.finish();

  console.log(`Evaluation complete: EM=${exactMatchRate.toFixed(3)}, F1=${avgF1.toFixed(3)}, IoU=${avgIoU.toFixed(3)}`);

  return {
    results,
    exactMatchRate,
    f1Score: avgF1,
    avgIoU,
    wandbRunUrl
  };
}

export const handler: Handler = async (event: EvaluationEvent | EventBridgeEvent<string, unknown> | APIGatewayProxyEvent) => {
  console.log('Evaluation runner invoked:', JSON.stringify(event, null, 2));

  // Parse event to get parameters
  let params: EvaluationEvent = {};

  if ('source' in event && event.source === 'aws.events') {
    // EventBridge scheduled event
    params = {
      modelName: 'Amazon Nova Pro',
      datasetVersion: 'latest',
      triggeredBy: 'EventBridge',
      triggerType: 'SCHEDULED'
    };
  } else if ('body' in event && event.body) {
    // API Gateway request
    params = JSON.parse(event.body);
    params.triggerType = 'MANUAL';
  } else {
    // Direct invocation
    params = event as EvaluationEvent;
  }

  const {
    jobId = `eval-${Date.now()}`,
    modelName = 'Amazon Nova Pro',
    datasetVersion = 'latest',
    triggeredBy = 'system',
    triggerType = 'MANUAL'
  } = params;

  try {
    // Update job status to RUNNING
    await updateJobStatus(jobId, 'RUNNING', {
      startedAt: new Date().toISOString(),
      modelName,
      datasetVersion,
      triggeredBy,
      triggerType
    });

    // Run evaluation
    const result = await runEvaluation(modelName, datasetVersion, jobId);

    // Update job status to COMPLETED
    await updateJobStatus(jobId, 'COMPLETED', {
      completedAt: new Date().toISOString(),
      exactMatchRate: result.exactMatchRate,
      f1Score: result.f1Score,
      avgIoU: result.avgIoU,
      sampleCount: result.results.length,
      wandbRunUrl: result.wandbRunUrl
    });

    return {
      statusCode: 200,
      body: JSON.stringify({
        success: true,
        jobId,
        exactMatchRate: result.exactMatchRate,
        f1Score: result.f1Score,
        avgIoU: result.avgIoU,
        sampleCount: result.results.length,
        wandbRunUrl: result.wandbRunUrl
      })
    };

  } catch (error) {
    console.error('Evaluation failed:', error);

    // Update job status to FAILED
    await updateJobStatus(jobId, 'FAILED', {
      completedAt: new Date().toISOString(),
      errorMessage: error instanceof Error ? error.message : 'Unknown error'
    });

    return {
      statusCode: 500,
      body: JSON.stringify({
        success: false,
        jobId,
        error: error instanceof Error ? error.message : 'Unknown error'
      })
    };
  }
};

File: application/amplify/functions/run-evaluation/handler.py
"""
Model Evaluation Lambda Handler

Evaluates AI models on the VQA dataset using ANLS and IoU metrics.
Triggered by SQS messages for parallel per-model evaluation.

Supports checkpoint/resume: when the Lambda approaches its timeout,
it saves progress to DynamoDB and re-enqueues the job to SQS so the
next invocation continues from where the previous one stopped.

Metrics:
  - ANLS (Average Normalized Levenshtein Similarity): text accuracy (DocVQA standard)
  - IoU (Intersection over Union): bounding box spatial accuracy
"""
import json
import os
import boto3
import wandb
from datetime import datetime, timezone
from decimal import Decimal
from typing import Dict, List, Optional
from huggingface_hub import snapshot_download
from PIL import Image
from io import BytesIO
from metrics import calculate_anls, calculate_iou
from prompts import get_evaluation_prompt

# AWS clients
dynamodb = boto3.resource('dynamodb')
bedrock_client = boto3.client('bedrock-runtime')
sqs_client = boto3.client('sqs')
ssm_client = boto3.client('ssm')

# Environment variables
WANDB_PROJECT = os.environ.get('WANDB_PROJECT', 'biz-doc-vqa')

# Time (ms) reserved before Lambda timeout for saving checkpoint and re-enqueueing.
# 120 seconds gives enough headroom for DynamoDB write + SQS send + W&B finish.
CHECKPOINT_BUFFER_MS = 120_000

# Module-level caches
_table_cache: Dict[str, object] = {}
_secrets_cache: Dict[str, str] = {}


# Table name overrides passed via SQS message from the data-stack trigger.
# This avoids circular dependency between function and data CloudFormation stacks.
_table_name_overrides: Dict[str, str] = {}

# Lambda context reference for timeout checking
_lambda_context = None


def get_table(prefix: str):
    """Get DynamoDB table by name from SQS message overrides."""
    if prefix not in _table_cache:
        table_name = _table_name_overrides.get(prefix)
        if not table_name:
            raise ValueError(f"Table name for '{prefix}' not provided in SQS message")
        print(f'Using table: {table_name}')
        _table_cache[prefix] = dynamodb.Table(table_name)
    return _table_cache[prefix]


def get_secret(ssm_param_env_key: str) -> str:
    """Read secret from SSM Parameter Store with caching.

    The environment variable contains the SSM parameter path (e.g. '/business-ocr/wandb-api-key').
    """
    ssm_param = os.environ.get(ssm_param_env_key, '')
    if not ssm_param:
        print(f'Warning: {ssm_param_env_key} environment variable is not set')
        return ''
    if ssm_param not in _secrets_cache:
        response = ssm_client.get_parameter(Name=ssm_param, WithDecryption=True)
        _secrets_cache[ssm_param] = response['Parameter']['Value']
    return _secrets_cache[ssm_param]


def handler(event, context):
    """
    SQS event handler for model evaluation.

    SQS message body format:
    {
        "evaluationJobId": "amplify-record-id",
        "jobId": "job-uuid",
        "datasetVersion": "v1.0.0",
        "huggingFaceRepoId": "icoxfog417/biz-doc-vqa",
        "modelId": "claude-sonnet-4-5",
        "modelName": "Claude Sonnet 4.5",
        "modelBedrockId": "global.anthropic.claude-sonnet-4-5-20250929-v1:0",
        "evaluationJobTableName": "EvaluationJob-xxx",
        "evaluationQueueUrl": "https://sqs.us-east-1.amazonaws.com/..."
    }
    """
    global _lambda_context
    _lambda_context = context

    batch_item_failures = []

    for record in event.get('Records', []):
        try:
            message = json.loads(record['body'])
            process_evaluation_job(message)
        except Exception as e:
            print(f"Error processing message {record.get('messageId')}: {str(e)}")
            batch_item_failures.append({'itemIdentifier': record['messageId']})

    return {'batchItemFailures': batch_item_failures}


def is_approaching_timeout() -> bool:
    """Check if the Lambda is approaching its timeout.

    Returns True when remaining execution time is less than CHECKPOINT_BUFFER_MS.
    """
    if _lambda_context is None:
        return False
    remaining_ms = _lambda_context.get_remaining_time_in_millis()
    return remaining_ms < CHECKPOINT_BUFFER_MS


def process_evaluation_job(message: Dict):
    """Process a single evaluation job for one model.

    Supports checkpoint/resume: if the previous invocation saved a checkpoint
    (due to approaching Lambda timeout), this invocation resumes from that point.
    The checkpoint stores the sample index, accumulated metrics, and counts in
    the DynamoDB EvaluationJob record.  When approaching timeout, the handler
    saves its progress, re-enqueues the job to SQS, and returns normally so
    the current SQS message is deleted (not counted as a retry failure).
    """
    evaluation_job_id = message['evaluationJobId']
    job_id = message['jobId']
    dataset_version = message['datasetVersion']
    hf_repo_id = message['huggingFaceRepoId']
    model_id = message['modelId']
    model_name = message.get('modelName', model_id)
    bedrock_model_id = message['modelBedrockId']
    queue_url = message.get('evaluationQueueUrl', '')

    # Table name is passed from the data-stack trigger to avoid circular
    # CloudFormation dependency between function and data stacks.
    eval_table_name = message.get('evaluationJobTableName')
    if not eval_table_name:
        raise ValueError('evaluationJobTableName not provided in SQS message')
    _table_name_overrides['EvaluationJob'] = eval_table_name

    job_table = get_table('EvaluationJob')
    now = datetime.now(timezone.utc).isoformat()

    # Read existing checkpoint from DynamoDB (if resuming)
    checkpoint = load_checkpoint(job_table, evaluation_job_id)
    start_index = checkpoint.get('sampleIndex', 0)
    is_resume = start_index > 0

    # Update status to RUNNING (only set startedAt on first invocation)
    if is_resume:
        print(f'Resuming from checkpoint at sample {start_index}')
        job_table.update_item(
            Key={'id': evaluation_job_id},
            UpdateExpression='SET #s = :status, updatedAt = :now',
            ExpressionAttributeNames={'#s': 'status'},
            ExpressionAttributeValues={':status': 'RUNNING', ':now': now},
        )
    else:
        job_table.update_item(
            Key={'id': evaluation_job_id},
            UpdateExpression='SET #s = :status, startedAt = :now, updatedAt = :now',
            ExpressionAttributeNames={'#s': 'status'},
            ExpressionAttributeValues={':status': 'RUNNING', ':now': now},
        )

    wandb_run = None

    try:
        # Initialize W&B
        wandb_api_key = get_secret('WANDB_API_KEY_SSM_PARAM')
        run_suffix = f'-part{checkpoint.get("part", 1)}' if is_resume else ''
        if wandb_api_key:
            wandb.login(key=wandb_api_key)
            wandb_run = wandb.init(
                project=WANDB_PROJECT,
                name=f'eval-{model_id}-{dataset_version}{run_suffix}',
                config={
                    'model_id': model_id,
                    'model_name': model_name,
                    'bedrock_model_id': bedrock_model_id,
                    'dataset_version': dataset_version,
                    'hf_repo_id': hf_repo_id,
                    'job_id': job_id,
                    'start_index': start_index,
                    'is_resume': is_resume,
                },
                tags=['evaluation', model_id, dataset_version],
            )

        # Download dataset from HF Hub (ImageFolder format)
        print(f'Downloading dataset from {hf_repo_id}...')
        hf_token = get_secret('HF_TOKEN_SSM_PARAM')
        dataset_dir = snapshot_download(
            hf_repo_id,
            repo_type='dataset',
            local_dir='/tmp/hf_dataset',
            token=hf_token if hf_token else None,
        )
        dataset = load_imagefolder_dataset(dataset_dir)
        total_samples = len(dataset)
        print(f'Loaded {total_samples} samples (starting from index {start_index})')

        # Restore accumulated metrics from checkpoint
        running_anls = checkpoint.get('runningAnls', 0.0)
        running_iou = checkpoint.get('runningIou', 0.0)
        samples_evaluated = checkpoint.get('samplesEvaluated', 0)
        samples_failed = checkpoint.get('samplesFailed', 0)
        failed_sample_errors: List[str] = []
        results_data: List[Dict] = []
        checkpointed = False

        for i in range(start_index, total_samples):
            # Check if we're approaching Lambda timeout
            if is_approaching_timeout():
                print(
                    f'Approaching timeout at sample {i}/{total_samples}. '
                    f'Saving checkpoint and re-enqueueing.'
                )
                save_checkpoint(
                    job_table, evaluation_job_id,
                    sample_index=i,
                    samples_evaluated=samples_evaluated,
                    samples_failed=samples_failed,
                    running_anls=running_anls,
                    running_iou=running_iou,
                    part=checkpoint.get('part', 1) + (1 if is_resume else 1),
                )
                reenqueue_job(message, queue_url)
                checkpointed = True
                break

            sample = dataset[i]
            try:
                image = Image.open(sample['_image_path'])
                prediction = invoke_model(
                    bedrock_model_id,
                    image,
                    sample['question'],
                    sample.get('language', 'en'),
                )

                # Calculate ANLS (text accuracy)
                anls = calculate_anls(prediction.get('answer', ''), sample['answers'])

                # Calculate IoU (bounding box accuracy)
                iou = calculate_iou(
                    prediction.get('bbox', [0.0, 0.0, 1.0, 1.0]),
                    list(sample['answer_bbox']),
                )

                samples_evaluated += 1
                # Incremental running average
                running_anls += (anls - running_anls) / samples_evaluated
                running_iou += (iou - running_iou) / samples_evaluated

                # Log incrementally to W&B
                if wandb_run:
                    wandb.log({
                        'progress/samples_evaluated': samples_evaluated,
                        'progress/samples_failed': samples_failed,
                        'progress/running_anls': running_anls,
                        'progress/running_iou': running_iou,
                        'sample/anls': anls,
                        'sample/iou': iou,
                    })

                results_data.append({
                    'annotation_id': sample.get('annotation_id', f'a{i}'),
                    'question': sample['question'],
                    'ground_truth': '\n'.join(sample['answers']) if sample['answers'] else '',
                    'prediction': prediction.get('answer', ''),
                    'anls': round(anls, 4),
                    'iou': round(iou, 4),
                })

                if samples_evaluated % 10 == 0:
                    print(
                        f'Progress: {samples_evaluated}/{total_samples} '
                        f'| ANLS: {running_anls:.4f} | IoU: {running_iou:.4f}'
                    )

            except Exception as e:
                samples_failed += 1
                error_msg = f'Sample {i}: {str(e)[:100]}'
                print(f'Error evaluating sample {i}: {str(e)}')
                # Keep first 10 error messages for diagnostics
                if len(failed_sample_errors) < 10:
                    failed_sample_errors.append(error_msg)
                continue

        # If we checkpointed, return early (job continues in next invocation)
        if checkpointed:
            print(
                f'Checkpoint saved for {model_id}: '
                f'evaluated={samples_evaluated}, failed={samples_failed}'
            )
            return

        # Log failure summary
        if samples_failed > 0:
            print(
                f'Warning: {samples_failed}/{total_samples} samples failed to evaluate. '
                f'First errors: {failed_sample_errors[:3]}'
            )

        # Log final results table to W&B
        wandb_run_url = ''
        if wandb_run and results_data:
            columns = ['annotation_id', 'question', 'ground_truth', 'prediction', 'anls', 'iou']
            table = wandb.Table(columns=columns)
            for row in results_data:
                table.add_data(*[row[c] for c in columns])
            wandb.log({'evaluation_results': table})

            wandb.summary['final_anls'] = running_anls
            wandb.summary['final_iou'] = running_iou
            wandb.summary['total_samples'] = samples_evaluated
            wandb.summary['failed_samples'] = samples_failed

            wandb_run_url = wandb_run.url or ''

        # Determine final status based on sample failures
        # COMPLETED: at least some samples evaluated successfully
        # FAILED: no samples evaluated (all failed)
        if samples_evaluated == 0 and total_samples > 0:
            raise RuntimeError(
                f'All {total_samples} samples failed to evaluate. '
                f'Errors: {"; ".join(failed_sample_errors[:3])}'
            )

        # Update job status to COMPLETED and clear checkpoint
        now = datetime.now(timezone.utc).isoformat()
        error_summary = (
            f'{samples_failed} samples failed: {"; ".join(failed_sample_errors[:3])}'
            if samples_failed > 0
            else None
        )
        update_expr = (
            'SET #s = :status, avgAnls = :anls, avgIou = :iou, '
            'totalSamples = :total, failedSamples = :failed, '
            'wandbRunUrl = :url, completedAt = :now, updatedAt = :now'
            + (', errorMessage = :error' if error_summary else '')
            + ' REMOVE checkpoint'
        )
        job_table.update_item(
            Key={'id': evaluation_job_id},
            UpdateExpression=update_expr,
            ExpressionAttributeNames={'#s': 'status'},
            ExpressionAttributeValues={
                ':status': 'COMPLETED',
                ':anls': Decimal(str(round(running_anls, 6))),
                ':iou': Decimal(str(round(running_iou, 6))),
                ':total': samples_evaluated,
                ':failed': samples_failed,
                ':url': wandb_run_url,
                ':now': now,
                **(({':error': error_summary}) if error_summary else {}),
            },
        )

        print(
            f'Evaluation complete for {model_id}: '
            f'ANLS={running_anls:.4f}, IoU={running_iou:.4f}, '
            f'Samples={samples_evaluated}, Failed={samples_failed}'
        )

    except Exception as e:
        print(f'Evaluation job failed for {model_id}: {str(e)}')
        now = datetime.now(timezone.utc).isoformat()
        job_table.update_item(
            Key={'id': evaluation_job_id},
            UpdateExpression='SET #s = :status, errorMessage = :error, updatedAt = :now',
            ExpressionAttributeNames={'#s': 'status'},
            ExpressionAttributeValues={
                ':status': 'FAILED',
                ':error': str(e)[:1000],
                ':now': now,
            },
        )
        raise  # Re-raise for SQS retry / DLQ

    finally:
        if wandb_run:
            wandb.finish()


def load_checkpoint(job_table, evaluation_job_id: str) -> Dict:
    """Load checkpoint data from DynamoDB EvaluationJob record.

    Returns checkpoint dict with keys: sampleIndex, samplesEvaluated,
    samplesFailed, runningAnls, runningIou, part.  Returns empty dict
    if no checkpoint exists.
    """
    response = job_table.get_item(
        Key={'id': evaluation_job_id},
        ProjectionExpression='checkpoint',
    )
    item = response.get('Item', {})
    checkpoint = item.get('checkpoint')
    if checkpoint:
        # Convert Decimal values back to float/int
        return {
            'sampleIndex': int(checkpoint.get('sampleIndex', 0)),
            'samplesEvaluated': int(checkpoint.get('samplesEvaluated', 0)),
            'samplesFailed': int(checkpoint.get('samplesFailed', 0)),
            'runningAnls': float(checkpoint.get('runningAnls', 0)),
            'runningIou': float(checkpoint.get('runningIou', 0)),
            'part': int(checkpoint.get('part', 1)),
        }
    return {}


def save_checkpoint(
    job_table,
    evaluation_job_id: str,
    sample_index: int,
    samples_evaluated: int,
    samples_failed: int,
    running_anls: float,
    running_iou: float,
    part: int,
):
    """Save checkpoint data to DynamoDB EvaluationJob record."""
    now = datetime.now(timezone.utc).isoformat()
    job_table.update_item(
        Key={'id': evaluation_job_id},
        UpdateExpression='SET checkpoint = :cp, updatedAt = :now',
        ExpressionAttributeValues={
            ':cp': {
                'sampleIndex': sample_index,
                'samplesEvaluated': samples_evaluated,
                'samplesFailed': samples_failed,
                'runningAnls': Decimal(str(round(running_anls, 8))),
                'runningIou': Decimal(str(round(running_iou, 8))),
                'part': part,
            },
            ':now': now,
        },
    )


def reenqueue_job(message: Dict, queue_url: str):
    """Re-enqueue the evaluation job to SQS to continue from checkpoint."""
    if not queue_url:
        print('Warning: evaluationQueueUrl not provided, cannot re-enqueue')
        return

    sqs_client.send_message(
        QueueUrl=queue_url,
        MessageBody=json.dumps(message),
    )
    print(f'Re-enqueued evaluation job {message["evaluationJobId"]} to continue')


def load_imagefolder_dataset(repo_dir: str) -> List[Dict]:
    """Load ImageFolder dataset from downloaded HF repo directory.

    Parses data/metadata.jsonl and returns list of samples with image file paths.
    """
    metadata_path = os.path.join(repo_dir, 'data', 'metadata.jsonl')
    if not os.path.exists(metadata_path):
        raise FileNotFoundError(f'metadata.jsonl not found at {metadata_path}')

    data_dir = os.path.join(repo_dir, 'data')
    samples: List[Dict] = []

    with open(metadata_path, 'r', encoding='utf-8') as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            row = json.loads(line)
            row['_image_path'] = os.path.join(data_dir, row['file_name'])
            samples.append(row)

    return samples


def invoke_model(bedrock_model_id: str, image, question: str, language: str) -> Dict:
    """Invoke Bedrock model with image and question, return parsed answer + bbox.

    Uses language-specific prompts synchronized with annotation prompts
    for consistent answer formatting (see prompts.py).
    """
    # Convert PIL Image to JPEG bytes
    img_buffer = BytesIO()
    if image.mode == 'RGBA':
        image = image.convert('RGB')
    image.save(img_buffer, format='JPEG', quality=85)
    img_bytes = img_buffer.getvalue()

    # Use language-specific prompt for consistent formatting with annotation
    prompt = get_evaluation_prompt(question, language)

    response = bedrock_client.converse(
        modelId=bedrock_model_id,
        messages=[
            {
                'role': 'user',
                'content': [
                    {'text': prompt},
                    {
                        'image': {
                            'format': 'jpeg',
                            'source': {'bytes': img_bytes},
                        },
                    },
                ],
            },
        ],
        inferenceConfig={
            'temperature': 0.1,
            'maxTokens': 500,
        },
    )

    content = response.get('output', {}).get('message', {}).get('content', [])
    response_text = content[0].get('text', '') if content else ''

    return parse_model_response(response_text)


def parse_model_response(response_text: str) -> Dict:
    """Parse model JSON response to extract answer and bounding box."""
    result: Dict = {'answer': '', 'bbox': [0.0, 0.0, 1.0, 1.0]}

    try:
        # Remove markdown code fences if present
        text = response_text.strip()
        if text.startswith('```'):
            lines = text.split('\n')
            text = '\n'.join(lines[1:])
            if text.endswith('```'):
                text = text[:-3]
            text = text.strip()

        parsed = json.loads(text)
        result['answer'] = str(parsed.get('answer', ''))

        bbox = parsed.get('bbox', [0, 0, 1, 1])
        if isinstance(bbox, list) and len(bbox) == 4:
            result['bbox'] = [float(v) for v in bbox]

    except (json.JSONDecodeError, ValueError):
        # Fallback: use raw text as answer
        result['answer'] = response_text.strip()

    return result


# Metrics functions (calculate_anls, calculate_iou) are imported from metrics.py
